### DISCUSSION TITLE
How to convert a package to MSL 4?

### DISCUSSION THREAD

**Discussion by bilderbuchi** [OpenModelica/OpenModelica#8726] [Q&A] [answered]
I am trying to get [XogenyTest ](https://github.com/xogeny/XogenyTest) to work with OM, and the first stumbling block is MSL 4 compatibility, so I thought I'd get that out of the way with the newly available(?) conversion script.

I load XogenyTest (warning about MSL version incompatibility shown), then run `OpenModelica.Scripting.convertPackageToLibrary`, whose docs say "Runs the conversion script for a library on a selected package"
```
OpenModelica.Scripting.convertPackageToLibrary(XogenyTest, Modelica, "4.0.0")
true
```
"true" means "success" apparently, but then I check in the folder where the library is stored (it's under Git), and the files are unchanged, and not converted. And indeed, unloading and reloading XogenyTest shows the same MSL version warning, and a spot check reveals MSL3 code (e.g. a ` Modelica.SIunits.Time`) is still present.

So,
* Why is the package not being converted? Am I using it wrong? 
* What does "success" mean if not a Package (files) Converted to Library Modelica 4?
* Should I be using another function/feature? I have not been able to find what/how to do this, instead.

FWIW, there is also the right-click item "convert to newer versions of used libraries", but that one confusingly wants to convert _from_ MSL 4 _to_ MSL 3, so I guess that does not work correctly, yet:
![image](https://user-images.githubusercontent.com/327442/159013389-0364c45c-6137-4b0c-9f5f-c2957beca4d7.png)

**Comment by sjoelund** [comment, marked as answer]
For me with XogenyTest in $HOME/.openmodelica/libraries, it updates the uses-annotation to 4.0.0 also on disk. I had the version installed though. Not sure that matters.

**Reply by bilderbuchi** [reply]
How did you "install" it? -- in the "Install libraries" dialog, it's not shown in the dropdown list of packages. I only have the repo cloned to disk (as I want to work on it) -- let's see what happens if I move it there manually.
(My MSL is in C:\OpenModelica1.19.0-dev-64bit\lib\omlibrary, not in the folder in APPDATA)

**Reply by bilderbuchi** [reply]
ah, I forgot to add more support level checkboxes ^^

**Reply by bilderbuchi** [reply]
Indeed, after "installing" the library I can run the conversion and the files change -- seems to be MSL updates from a couple spot checks, but also formatting was changed.

I'll just copy those over into my local clone that is not installed, so for now that is good enough. I'm glad that I don't have a user package to convert to, as that *would not work*.

Thanks!

**Reply by bilderbuchi** [reply]
BTW, is it intentional that there is _both_ an `openmodelica` and a  `.openmodelica` folder in `AppData\Roaming`, or is this from using a nightly in some transition period? Feels weird to me. I was looking in the wrong one and curious where that libraries folder went? :D

**Reply by sjoelund** [reply]
@adeas31 it seems OMEdit only issues save commands; not using the diff algorithm for converted libraries. Any reason it's different from saving a library manually?

**Reply by sjoelund** [reply]
On Linux there is $HOME/.openmodelica and $HOME/.config/openmodelica/omedit.ini (the latter is chosen by Qt)

**Reply by adeas31** [reply]
> @adeas31 it seems OMEdit only issues save commands; not using the diff algorithm for converted libraries. Any reason it's different from saving a library manually?

diff is only needed to preserve the formatting. I don't want to run the diff on the whole library and make the user wait forever and show them hundred of error messages (yes diff algorithm fails too often).

**Reply by adeas31** [reply]
> On Linux there is $HOME/.openmodelica and $HOME/.config/openmodelica/omedit.ini (the latter is chosen by Qt)

I think the confusion is that OMEdit settings are saved in `openmodelica` and libraries are saved in `.openmodelica`. See also #5931.
I propose that we should use `$HOME/openmodelica` (on Windows) and `$HOME/.config/openmodelica/` (on Linux) even for libraries. Package manager is relatively new and not alot of users has started using it. So changing the libraries location is safer option in comparison to changing the OMEdit settings location.

**Comment by perost** [comment]
`convertPackageToLibrary` only converts a package loaded by the compiler, it doesn't also save it to file. So if you use the scripting API you first need to convert the library and then save it, though I'm not sure how to do that. Using OMEdit is probably easier if it works correctly.

But as you say it seems like the "From" and "To" in OMEdit have been reversed, though it seems like it's only a graphical issue and the actual conversion will be from 3.2.2 to 4.0.0.

The conversion in OMEdit seems to have some issues with user models though. I made a new class and added a uses-annotation to it, and after the conversion I couldn't save the class because OMEdit considered it to be a system library. @adeas31 might need to have a look at that.

**Reply by bilderbuchi** [reply]
Thanks, that gave the necessary hints - I "installed" the package (luckily available to) and ignored the wrong "direction", got it converted (and will copy the files over manually).

As you say, though, this would not work for user models (that you cannot *install* via OMEdit).

**Reply by adeas31** [reply]
> But as you say it seems like the "From" and "To" in OMEdit have been reversed, though it seems like it's only a graphical issue and the actual conversion will be from 3.2.2 to 4.0.0.

See PR #8730

> 
> The conversion in OMEdit seems to have some issues with user models though. I made a new class and added a uses-annotation to it, and after the conversion I couldn't save the class because OMEdit considered it to be a system library. @adeas31 might need to have a look at that.

Please share a MWE.

**Comment by casella** [comment]
@bilderbuchi, I'll look at this ASAP. The idea is that one should convert _his own_ libraries, not the read-only installed system libraries. If that doesn't work, we need to fix it.

**Reply by bilderbuchi** [reply]
Thanks! FYI, there's no urgency on this from my side, as [outlined above](https://github.com/OpenModelica/OpenModelica/discussions/8726#discussioncomment-2388060), I worked around the issue for now, and luckily I'm not converting packages every week. :D

**Comment by casella** [comment]
@adeas31, here is as simple MWE.
```
package TestConversion
  model M
    Modelica.SIunits.Length L = 1;
  end M;
  annotation(uses(Modelica(version="3.2.3")));
end TestConversion;
```
I just loaded it with v1.19.0-dev-697-gfa7189b918, note that by default I'm not loading any version of MSL. OMEdit automatically loads MSL 3.2.3. If I now right-click on TestConversion, I get this pop-up window
![immagine](https://user-images.githubusercontent.com/4233148/159373803-1f99e00e-60fa-4bf6-8b17-fee8734ea91b.png)
which has at least two problems:
1. The from and to column are apparently flipped
2. It tells me MSL 4.0.0 is not installed, while it is. Maybe that's because I did not install it with the package manager, but rather had the OpenModelica installer put it in the installation directory?

I'm not sure if the latest commits by @adeas31 have already fixed this issue. @adeas31 can you please comment on that?

**Comment by adeas31** [comment]
I made following fixes yesterday,

- #8730
- #8733
- #8737 

> It tells me MSL 4.0.0 is not installed, while it is. Maybe that's because I did not install it with the package manager, but rather had the OpenModelica installer put it in the installation directory?

It uses `getAvailableLibraryVersions()` to check which versions of the library are installed. I checked it again and it works as expected. If the users selects the not installed version during the conversion then we anyway install it first and then perform the conversion. So perhaps we should remove this information as is confusing for the user.

**Reply by bilderbuchi** [reply]
> as is confusing for the user.

Indeed, the list claimed that none of the MSL options were installed, while I had MSL components open, but I just raised my eyebrow and moved on 😌

**Reply by adeas31** [reply]
PR #8741 is on the way.
